File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-0.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        lastSeen: Dict[str, int] = {}

        maxLen = 0
        left = 0
        for right, ch in enumerate(s):
            if ch in lastSeen:
                left = max(left, lastSeen.get(ch, 0) + 1)
            lastSeen[ch] = right

            maxLen = max(maxLen, right - left + 1)

        return maxLen

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-0.java
class Solution {
    public int characterReplacement(String s, int k) {
        int left = 0;
        int best = 0;
        int maxCount = 0;

        int[] counter = new int[26];
        for (int right = 0; right < s.length(); right++) {
            counter[s.charAt(right) - 'A'] += 1;
            maxCount = Math.max(maxCount, counter[s.charAt(right) - 'A']);
            while ((right-left + 1) - maxCount > k) {
                counter[s.charAt(left) - 'A'] -= 1;
                left++;
            }

            best = Math.max(best, right - left + 1);
        } 

        return best;
    }
}
